File: solutions/python/protein-translation/1/protein_translation.py
protein_trans_table: dict[str, str] = {
    'AUG': 'Methionine',
    'UUU': 'Phenylalanine',
    'UUC': 'Phenylalanine',
    'UUA': 'Leucine',
    'UUG': 'Leucine',
    'UCU': 'Serine',
    'UCC': 'Serine',
    'UCA': 'Serine',
    'UCG': 'Serine',
    'UAU': 'Tyrosine',
    'UAC': 'Tyrosine',
    'UGU': 'Cysteine',
    'UGC': 'Cysteine',
    'UGG': 'Tryptophan',
    'UAA': 'STOP',
    'UAG': 'STOP',
    'UGA': 'STOP'
}


def proteins(strand: str):
    group_len = 3
    sequence = [
        protein_trans_table.get(strand[i:i+group_len])
        for i in range(0, len(strand), group_len)
    ]
    try:
        index = sequence.index('STOP')
        sequence = sequence[:index]
    except ValueError:
        pass

    return sequence

File: solutions/ruby/protein-translation/1/protein_translation.rb
class InvalidCodonError < StandardError
  def initialize(message)
    message
  end
end

module Translation
  PROTEIN = {
    'AUG' => "Methionine",
    **['UUU', 'UUC'].map{|codon| [codon, "Phenylalanine"]}.to_h,
    **['UUA', 'UUG'].map{|codon| [codon, "Leucine"]}.to_h,
    **['UCU', 'UCC', 'UCA', 'UCG'].map{|codon| [codon, "Serine"]}.to_h,
    **['UAU', 'UAC'].map{|codon| [codon, "Tyrosine"]}.to_h,
    **['UGU', 'UGC'].map{|codon| [codon, "Cysteine"]}.to_h,
    'UGG' => 'Tryptophan',
    **['UAA', 'UAG', 'UGA'].map{|codon| [codon, :stop]}.to_h
  }

  def self.of_rna(strand)
    codons(strand)
      .lazy
      .map{|codon|
        protein = PROTEIN[codon]
        raise InvalidCodonError.new(protein) if protein.nil?
        protein
      }
      .take_while {|protein| protein != :stop}
      .force
  end

  def self.codons(strand)
    strand.chars.each_slice(3).to_a.map(&:join)
  end
end
